Bug 2087037: Rebase onto latest master from upstream

JoelSpeed opened the pull request:
This is a rebase of the Autoscaler onto the master branch of the upstream repository. This ensures we have the latest changes ahead of the next release.

I used the rebase bot to create the branch initially but then made some changes afterwards.

I rebased onto kubernetes/master and squashed all automated ART commits into the setup commit, as well as dropping an additional owners file for cherry-server that was need. I've then gone through each of the other commits to make sure they are clean and resolve any conflicts that exist.

I did have to add a merge commit to make this pass, but, that merge commit is clean (which I think we should strive for in the future).

Additionally, I picked the vendor update for the Kube 1.24 branch, though I've set this as a drop so that we can pick the 1.25 equivalent next time.

enxebre commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller_test.go:
```diff
@@ -73,6 +73,14 @@ const customCAPIGroup = "custom.x-k8s.io"
 const fifteenSecondDuration = time.Second * 15
```
do we really need to carry this commit? can't this test case be covered upstream?

JoelSpeed replied:
I think you're right, apart from the changes in [clusterapi_utils_test.go](https://github.com/openshift/kubernetes-autoscaler/pull/231/commits/5c3ac70ed793f921453516ab6bad0db0b9f1c82a#diff-44f848f51d0ba33c7696002d371d4549b0daa7d8e6dbf173cfda12b8f125b0c9), we are not really adding anything. We will still need that small fix for the clusterapi_utils test though otherwise our other changes which set the group to machine.openshift.io break the test

enxebre commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup.go:
```diff
@@ -33,7 +33,9 @@ import (
 
 const (
-	debugFormat                    = "%s (min: %d, max: %d, replicas: %d)"
+	capiMachineDeleteAnnotationKey = "cluster.x-k8s.io/delete-machine"
```
please don't instantiate this annotations here, just use the same code than upstream https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/clusterapi/clusterapi_utils.go#L56-L68

The only thing that needs to be instantiated statically is the delete annotation for mapi, see https://github.com/openshift/kubernetes-autoscaler/pull/230/files#diff-8266b9d377a9df07645393a09139352c01636b7c84b6678586245e72c69dfe45R38

enxebre replied:
To clarify, I'm not proposing to switch over anything right now. All we need is now upstream which dynamically sets the group in the annotations via the get*annotationKey. https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/clusterapi/clusterapi_utils.go#L56-L68

So we need to drop everything else. The only skew we need is to have static `oldMachineDeleteAnnotationKey = "machine.openshift.io/cluster-api-delete-machine" because it's what mao uses and has a different suffix (so the dynamic get*annotation won't work for mapi in this case.).

Now, I'd suggest to eventually update mao to migrate the annotation to match upstream suffix (delete-machine) so eventually you don't need to carry this static annotation here.

JoelSpeed commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_utils.go:
```diff
@@ -29,6 +29,10 @@ import (
 )
 
 const (
+	capiNodeGroupMinSizeAnnotationKey = "cluster.x-k8s.io/cluster-api-autoscaler-node-group-min-size"
```
This is needed for the fallback nature of this commit. The addition here is so that we can check for the CAPI group if the MAPI group doesn't exist. Using `nodeGroupMinSizeAnnotationKey` will only return the MAPI version in a MAPI based cluster.

If we are dropping these, then we may as well just drop this commit entirely (which, looking at the reason it was originally added, may actually be suitable)

JoelSpeed commented:
@enxebre I've dropped those test changes as you were right, they don't seem to be necessary anymore. I've also added the handling of the old machine delete key in a separate commit, PTAL

JoelSpeed commented:
I've prepped a patch for MAO to migrate to the new annotation format, https://github.com/openshift/machine-api-operator/pull/1024

enxebre commented:
thanks!
/lgtm

enxebre commented:
the last commit you added I guess we'll want to conflate it with this one [here](https://github.com/openshift/kubernetes-autoscaler/pull/231/commits/f2c779b9bd5d00e5145fab0ed288c3a5983ede15#diff-ee9aff87bf5264792f8f8ac23b959ff38ad79096ca0c949c535f01350e104f2d)

/lgtm

JoelSpeed commented:
> the last commit you added I guess we'll want to conflate it with this one [here](https://github.com/openshift/kubernetes-autoscaler/pull/231/commits/f2c779b9bd5d00e5145fab0ed288c3a5983ede15#diff-ee9aff87bf5264792f8f8ac23b959ff38ad79096ca0c949c535f01350e104f2d)

Yep, just wanted to check the images job passes, if it does, I will rebase and merge that in to clean up the history

enxebre commented:
/lgtm

elmiko commented:
the ordering of our carry commits looks good, and the testing errors seem to be flakes. i'm good to merge this but do we need to do anything special with regards to #229 ?

/approve

JoelSpeed commented:
/approve

Not sure why Mike's label didn't apply

JoelSpeed commented:
/retest-required

enxebre commented:
/retest

elmiko commented:
the failures on the e2e-aws-operator test seem to be related to network failures or connection issues with specific pods. i still do not believe this is related to the autoscaler changes, but i'm not sure why they keep failing with different results.

as for the github action job, i do not believe we should have these enabled for our fork of this repo. this requires a little more investigation to figure out what we should be doing here.

/test e2e-aws-operator

JoelSpeed commented:
/test e2e-aws-operator

JoelSpeed commented:
/override test-and-verify

We shouldn't be using the actions, all openshift CI has passed

JoelSpeed commented:
@elmiko and I had a chat, we are going to merge this PR as is and follow up to resolve the test-and-verify issue which seems to have been introduced by some stricter linting on the upstream repository. Expect a follow up shortly
